### DISCUSSION TITLE
Strange compilation error CS0407 even type constraints are specified

### DISCUSSION THREAD

**Discussion by rvhuang** [dotnet/csharplang#1573] [Q&A] [answered]
Not sure if the question is suitable here.
Consider following scenario:

```csharp
using System; 

namespace ConsoleApp1
{
    class Program
    {
        static T GetInstance<T>()
            where T : IDisposable, new()
        {
            return new T();
        }

        static void DoSomethingOverload1<T>()
            where T : IDisposable, new()
        {
            // Error CS0407
            // 'T Program.GetInstance<T>()' has the wrong return type
            DoSomething(GetInstance<T>);
        }

        static void DoSomethingOverload2<T>()
            where T : IDisposable, new()
        {
            // No error
            DoSomething(() => new T());
        }

        static void DoSomething(Func<IDisposable> factory)  
        {
            // TODO
        }
    }
}

```
When `DoSomethingOverload1()` tries to call `DoSomething()`, the compiler does not consider that `GetInstance<T>()` has valid return type. However, both `DoSomethingOverload1()` and `GetInstance<T>()` explicitly specify that `T` should implement `IDisposable` which should fit the signature of `DoSomething()`. The error does not occur in `DoSomethingOverload2()`. 

Thanks for reading the question.

**Comment by ufcpp** [comment, marked as answer]
```cs
        static void DoSomethingOverload1<T>()
            where T : class, IDisposable, new()
        {
            // No error
            DoSomething(GetInstance<T>);
        }
```

Delegate variance requires a reference type. So you need `class` constraint.

**Comment by ethanae** [comment]
@rvhuang, more of a StackOverflow question. Proposal to have this thread closed. Although insightful it doesn't address issues within the language.

**Comment by iam3yal** [comment]
@rvhuang in addition to what @ethanae said, if you're a bit intimidated by SO or don't use it then you can also  create a [gist](https://gist.github.com/) for the code and then ask questions at [C# channel](https://gitter.im/dotnet/csharplang), alternatively, join the community that fits your needs, check it out [here](https://github.com/dotnet/csharplang/issues/1145).
